libmamba/src/core/download.cpp

```diff
@@ -0,0 +1,609 @@
+
+        p_handle->configure_handle(
+            p_request->url,
+            set_low_speed_opt,
+            context.remote_fetch_params.connect_timeout_secs,
+            set_ssl_no_revoke,
+            proxy_match(p_request->url),
+            context.remote_fetch_params.ssl_verify
+        );
+
+        p_handle->set_opt(CURLOPT_HEADERFUNCTION, &DownloadAttempt::curl_header_callback);
+        p_handle->set_opt(CURLOPT_HEADERDATA, this);
+
+        p_handle->set_opt(CURLOPT_WRITEFUNCTION, &DownloadAttempt::curl_write_callback);
+        p_handle->set_opt(CURLOPT_WRITEDATA, this);
+
+        if (p_request->progress.has_value())
+        {
+            p_handle->set_opt(CURLOPT_XFERINFOFUNCTION, &DownloadAttempt::curl_progress_callback);
+            p_handle->set_opt(CURLOPT_XFERINFODATA, this);
+            p_handle->set_opt(CURLOPT_NOPROGRESS, 0L);
+        }
+
+        if (ends_with(p_request->url, ".json"))
+        {
+            // accept all encodings supported by the libcurl build
+            p_handle->set_opt(CURLOPT_ACCEPT_ENCODING, "");
+            p_handle->add_header("Content-Type: application/json");
+        }
+
+        p_handle->set_opt(CURLOPT_VERBOSE, context.output_params.verbosity >= 2);
+        
+        configure_handle_headers(context);
+
+        auto logger = spdlog::get("libcurl");
+        p_handle->set_opt(CURLOPT_DEBUGFUNCTION, curl_debug_callback);
+        p_handle->set_opt(CURLOPT_DEBUGDATA, logger.get());
+    }
+
+    void DownloadAttempt::configure_handle_headers(const Context& context)
+    {
+        p_handle->reset_headers();
+
+        std::string user_agent = fmt::format(
+            "User-Agent: {} {}",
+            context.remote_fetch_params.user_agent,
+            curl_version()
+        );
+        p_handle->add_header(user_agent);
+
+        // get url host
+        const auto url_handler = URLHandler(p_request->url);
+        auto host = url_handler.host();
+        const auto port = url_handler.port();
+        if (port.size())
+        {
+            host += ":" + port;
+        }
+
+        if (context.authentication_info().count(host))
+        {
+            const auto& auth = context.authentication_info().at(host);
+            if (auth.type == AuthenticationType::kBearerToken)
+            {
+                p_handle->add_header(fmt::format("Authorization: Bearer {}", auth.value));
+            }
+        }
+        
+        if (p_request->if_none_match.has_value())
+        {
+            p_handle->add_header("If-None-Match:" + p_request->if_none_match.value());
+        }
+
+        if (p_request->if_modified_since.has_value())
+        {
+            p_handle->add_header("If-Modified-Since:" + p_request->if_modified_since.value());
+        }
+       
+        p_handle->set_opt_header();
+    }
+
+    size_t DownloadAttempt::write_data(char* buffer, size_t size)
+    {
+        if (!m_file.is_open())
```

Comment:
At this point `m_file` needs to be set right? Maybe add an assertion?

Reply:
Not necessarily, here we rely on lazy opening